src/core/sock/sockinfo_tcp.cpp

```diff
@@ -2269,6 +2354,123 @@ int sockinfo_tcp::connect(const sockaddr *__to, socklen_t __tolen)
     return 0;
 }
 
+bool sockinfo_tcp::connect_bind_any_and_check_rules()
+{
+    bool bound_any_addr = m_bound.is_anyaddr();
+    if (bound_any_addr) {
+        const ip_address &ip = m_p_connected_dst_entry->get_src_addr();
+        // The family of local_addr may change due to mapped IPv4.
+        m_bound.set_ip_port(m_p_connected_dst_entry->get_sa_family(), &ip, m_bound.get_in_port());
+    }
+
+    IF_STATS(m_p_socket_stats->set_bound_if(m_bound));
+
+    sock_addr remote_addr;
+    remote_addr.set_sa_family(m_p_connected_dst_entry->get_sa_family());
+    remote_addr.set_in_addr(m_p_connected_dst_entry->get_dst_addr());
+    remote_addr.set_in_port(m_p_connected_dst_entry->get_dst_port());
+    if (!m_p_connected_dst_entry->is_offloaded() ||
+        find_target_family(ROLE_TCP_CLIENT, (sockaddr *)&remote_addr, m_bound.get_p_sa()) !=
+            TRANS_XLIO) {
+        passthrough_unlock("non offloaded socket --> connect only via OS");
+        return false;
+    } else if (has_epoll_context()) {
+        m_econtext->remove_fd_from_epoll_os(m_fd); // remove fd from os epoll
+    }
+
+    if (bound_any_addr) {
+        tcp_bind(&m_pcb, reinterpret_cast<const ip_addr_t *>(&m_bound.get_ip_addr()),
+                 ntohs(m_bound.get_in_port()), m_pcb.is_ipv6);
+    }
+
+    m_conn_state = TCP_CONN_CONNECTING;
+    return true;
+}
+
+int sockinfo_tcp::connect_async_return()
+{
+    errno = EINPROGRESS;
+    m_error_status = EINPROGRESS;
+    m_sock_state = TCP_SOCK_ASYNC_CONNECT;
+    unlock_tcp_con();
```

Comment:
I'd prefer the entity which acquires a lock to release it. Asymmetric unlock is error prone and less readable. This can be refactored later.

Reply:
connect_async_return() is called from the app thread and is just a helper method of moved code to avoid duplicate code. Since its is used also in the old path i was concerned about playing with the lock location. I dont't like it either. i'll see what can be done.